sql/tests: TestRandomSyntaxSelect failed

sql/tests.TestRandomSyntaxSelect [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2973750&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2973750&tab=artifacts#/) on master @ [88754f2866fde1f37620b04c6a708eefc0b5be54](https://github.com/cockroachdb/cockroach/commits/88754f2866fde1f37620b04c6a708eefc0b5be54):


```
    rsg_test.go:731: 1m20s of 5m0s: 9197 executions, 197 successful
    rsg_test.go:731: 1m25s of 5m0s: 9685 executions, 211 successful
    rsg_test.go:731: 1m30s of 5m0s: 10183 executions, 224 successful
    rsg_test.go:731: 1m35s of 5m0s: 10694 executions, 237 successful
    rsg_test.go:731: 1m40s of 5m0s: 11181 executions, 246 successful
    rsg_test.go:731: 1m45s of 5m0s: 11643 executions, 255 successful
    rsg_test.go:731: 1m50s of 5m0s: 12127 executions, 269 successful
    rsg_test.go:731: 1m55s of 5m0s: 12623 executions, 285 successful
    rsg_test.go:731: 2m0s of 5m0s: 13094 executions, 294 successful
    rsg_test.go:731: 2m5s of 5m0s: 13565 executions, 307 successful
    rsg_test.go:731: 2m10s of 5m0s: 14060 executions, 324 successful
    rsg_test.go:731: 2m15s of 5m0s: 14530 executions, 334 successful
    rsg_test.go:731: 2m20s of 5m0s: 15016 executions, 342 successful
    rsg_test.go:731: 2m25s of 5m0s: 15497 executions, 354 successful
    rsg_test.go:731: 2m30s of 5m0s: 15944 executions, 361 successful
    rsg_test.go:731: 2m35s of 5m0s: 16424 executions, 377 successful
    rsg_test.go:731: 2m40s of 5m0s: 16915 executions, 386 successful
    rsg_test.go:731: 2m45s of 5m0s: 17352 executions, 395 successful
    rsg_test.go:731: 2m50s of 5m0s: 17810 executions, 403 successful
    rsg_test.go:731: 2m55s of 5m0s: 18290 executions, 408 successful
    rsg_test.go:731: 3m0s of 5m0s: 18765 executions, 416 successful
    rsg_test.go:731: 3m5s of 5m0s: 19191 executions, 426 successful
    rsg_test.go:731: 3m10s of 5m0s: 19634 executions, 435 successful
    rsg_test.go:731: 3m15s of 5m0s: 20078 executions, 448 successful
    rsg_test.go:731: 3m20s of 5m0s: 20523 executions, 453 successful
    rsg_test.go:731: 3m25s of 5m0s: 20986 executions, 460 successful
    rsg_test.go:731: 3m30s of 5m0s: 21471 executions, 468 successful
    rsg_test.go:731: 3m35s of 5m0s: 21952 executions, 478 successful
    rsg_test.go:731: 3m40s of 5m0s: 22392 executions, 484 successful
    rsg_test.go:731: 3m45s of 5m0s: 22825 executions, 492 successful
    rsg_test.go:731: 3m50s of 5m0s: 23265 executions, 505 successful
    rsg_test.go:731: 3m55s of 5m0s: 23703 executions, 516 successful
    rsg_test.go:731: 4m0s of 5m0s: 24148 executions, 521 successful
    rsg_test.go:731: 4m5s of 5m0s: 24596 executions, 530 successful
    rsg_test.go:731: 4m10s of 5m0s: 25044 executions, 544 successful
    rsg_test.go:731: 4m15s of 5m0s: 25500 executions, 552 successful
    rsg_test.go:731: 4m20s of 5m0s: 25935 executions, 560 successful
    rsg_test.go:731: 4m25s of 5m0s: 26365 executions, 570 successful
    rsg_test.go:731: 4m30s of 5m0s: 26816 executions, 583 successful
    rsg_test.go:731: 4m35s of 5m0s: 27252 executions, 594 successful
    rsg_test.go:731: 4m40s of 5m0s: 27699 executions, 604 successful
    rsg_test.go:731: 4m45s of 5m0s: 28175 executions, 615 successful
    rsg_test.go:731: 4m50s of 5m0s: 28610 executions, 623 successful
    rsg_test.go:731: 4m55s of 5m0s: 29017 executions, 630 successful
    rsg_test.go:731: 5m0s of 5m0s: 29425 executions, 637 successful
    rsg_test.go:767: 29433 executions, 637 successful
    rsg_test.go:775: Parse followed by Format is not idempotent: "SELECT OPERATOR ( ~ ) DEFAULT COLLATE ident AS ILIKE FROM ident " -> "SELECT ~DEFAULT COLLATE ident AS ilike FROM ident" != "SELECT (~DEFAULT) COLLATE ident AS ilike FROM ident"
    rsg_test.go:255: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxSelect141229647
--- FAIL: TestRandomSyntaxSelect (300.43s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxSelect PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #64396 sql/tests: TestRandomSyntaxSelect failed [C-test-failure O-robot branch-release-21.1]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-queries @rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSelect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxSelect [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2985382&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2985382&tab=artifacts#/) on master @ [b706c913c07890d04d6bbec4d9d7cd8fea609b95](https://github.com/cockroachdb/cockroach/commits/b706c913c07890d04d6bbec4d9d7cd8fea609b95):


```
    rsg_test.go:731: 1m20s of 5m0s: 9794 executions, 223 successful
    rsg_test.go:731: 1m25s of 5m0s: 10319 executions, 236 successful
    rsg_test.go:731: 1m30s of 5m0s: 10817 executions, 251 successful
    rsg_test.go:731: 1m35s of 5m0s: 11278 executions, 266 successful
    rsg_test.go:731: 1m40s of 5m0s: 11826 executions, 281 successful
    rsg_test.go:731: 1m45s of 5m0s: 12316 executions, 287 successful
    rsg_test.go:731: 1m50s of 5m0s: 12797 executions, 304 successful
    rsg_test.go:731: 1m55s of 5m0s: 13304 executions, 321 successful
    rsg_test.go:731: 2m0s of 5m0s: 13803 executions, 329 successful
    rsg_test.go:731: 2m5s of 5m0s: 14293 executions, 343 successful
    rsg_test.go:731: 2m10s of 5m0s: 14807 executions, 353 successful
    rsg_test.go:731: 2m15s of 5m0s: 15288 executions, 370 successful
    rsg_test.go:731: 2m20s of 5m0s: 15786 executions, 381 successful
    rsg_test.go:731: 2m25s of 5m0s: 16294 executions, 401 successful
    rsg_test.go:731: 2m30s of 5m0s: 16786 executions, 409 successful
    rsg_test.go:731: 2m35s of 5m0s: 17255 executions, 420 successful
    rsg_test.go:731: 2m40s of 5m0s: 17727 executions, 428 successful
    rsg_test.go:731: 2m45s of 5m0s: 18248 executions, 439 successful
    rsg_test.go:731: 2m50s of 5m0s: 18718 executions, 451 successful
    rsg_test.go:731: 2m55s of 5m0s: 19195 executions, 461 successful
    rsg_test.go:731: 3m0s of 5m0s: 19650 executions, 475 successful
    rsg_test.go:731: 3m5s of 5m0s: 20112 executions, 485 successful
    rsg_test.go:731: 3m10s of 5m0s: 20574 executions, 492 successful
    rsg_test.go:731: 3m15s of 5m0s: 21026 executions, 504 successful
    rsg_test.go:731: 3m20s of 5m0s: 21491 executions, 516 successful
    rsg_test.go:731: 3m25s of 5m0s: 21983 executions, 521 successful
    rsg_test.go:731: 3m30s of 5m0s: 22451 executions, 531 successful
    rsg_test.go:731: 3m35s of 5m0s: 22893 executions, 538 successful
    rsg_test.go:731: 3m40s of 5m0s: 23367 executions, 548 successful
    rsg_test.go:731: 3m45s of 5m0s: 23820 executions, 555 successful
    rsg_test.go:731: 3m50s of 5m0s: 24265 executions, 562 successful
    rsg_test.go:731: 3m55s of 5m0s: 24735 executions, 569 successful
    rsg_test.go:731: 4m0s of 5m0s: 25230 executions, 582 successful
    rsg_test.go:731: 4m5s of 5m0s: 25683 executions, 594 successful
    rsg_test.go:731: 4m10s of 5m0s: 26162 executions, 610 successful
    rsg_test.go:731: 4m15s of 5m0s: 26626 executions, 621 successful
    rsg_test.go:731: 4m20s of 5m0s: 27098 executions, 632 successful
    rsg_test.go:731: 4m25s of 5m0s: 27544 executions, 644 successful
    rsg_test.go:731: 4m30s of 5m0s: 27998 executions, 652 successful
    rsg_test.go:731: 4m35s of 5m0s: 28439 executions, 661 successful
    rsg_test.go:731: 4m40s of 5m0s: 28908 executions, 676 successful
    rsg_test.go:731: 4m45s of 5m0s: 29354 executions, 682 successful
    rsg_test.go:731: 4m50s of 5m0s: 29774 executions, 687 successful
    rsg_test.go:731: 4m55s of 5m0s: 30252 executions, 698 successful
    rsg_test.go:731: 5m0s of 5m0s: 30700 executions, 705 successful
    rsg_test.go:767: 30708 executions, 705 successful
    rsg_test.go:775: Parse followed by Format is not idempotent: "SELECT OPERATOR ( - ) ident . VALUES . ident . * COLLATE SYMMETRIC ident FROM ident " -> "SELECT -ident.values.ident.* COLLATE symmetric AS ident FROM ident" != "SELECT (-ident.values.ident.*) COLLATE symmetric AS ident FROM ident"
    rsg_test.go:255: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxSelect040496097
--- FAIL: TestRandomSyntaxSelect (300.35s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxSelect PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #64396 sql/tests: TestRandomSyntaxSelect failed [C-test-failure O-robot branch-release-21.1]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-queries @rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSelect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/opt/exec/execbuilder/scalar.go
+++ b/pkg/sql/opt/exec/execbuilder/scalar.go
@@ -250,7 +250,7 @@ func (b *Builder) buildUnary(ctx *buildScalarCtx, scalar opt.ScalarExpr) (tree.T
 		return nil, err
 	}
 	operator := opt.UnaryOpReverseMap[scalar.Op()]
-	return tree.NewTypedUnaryExpr(operator, input, scalar.DataType()), nil
+	return tree.NewTypedUnaryExpr(tree.MakeUnaryOperator(operator), input, scalar.DataType()), nil
 }
 
 func (b *Builder) buildBinary(ctx *buildScalarCtx, scalar opt.ScalarExpr) (tree.TypedExpr, error) {

--- a/pkg/sql/opt/operator.go
+++ b/pkg/sql/opt/operator.go
@@ -169,7 +169,7 @@ var BinaryOpReverseMap = map[Operator]tree.BinaryOperator{
 
 // UnaryOpReverseMap maps from an optimizer operator type to a semantic tree
 // unary operator type.
-var UnaryOpReverseMap = map[Operator]tree.UnaryOperator{
+var UnaryOpReverseMap = map[Operator]tree.UnaryOperatorSymbol{
 	UnaryMinusOp:      tree.UnaryMinus,
 	UnaryComplementOp: tree.UnaryComplement,
 	UnarySqrtOp:       tree.UnarySqrt,

--- a/pkg/sql/opt/optbuilder/scalar.go
+++ b/pkg/sql/opt/optbuilder/scalar.go
@@ -765,7 +765,7 @@ func (b *Builder) constructBinary(
 func (b *Builder) constructUnary(
 	un tree.UnaryOperator, input opt.ScalarExpr, typ *types.T,
 ) opt.ScalarExpr {
-	switch un {
+	switch un.Symbol {
 	case tree.UnaryMinus:
 		return b.factory.ConstructUnaryMinus(input)
 	case tree.UnaryComplement:

--- a/pkg/sql/parser/parse.go
+++ b/pkg/sql/parser/parse.go
@@ -228,7 +228,10 @@ func unaryNegation(e tree.Expr) tree.Expr {
 	}
 
 	// Common case.
-	return &tree.UnaryExpr{Operator: tree.UnaryMinus, Expr: e}
+	return &tree.UnaryExpr{
+		Operator: tree.MakeUnaryOperator(tree.UnaryMinus),
+		Expr:     e,
+	}
 }
 
 // Parse parses a sql statement string and returns a list of Statements.

--- a/pkg/sql/parser/parse_test.go
+++ b/pkg/sql/parser/parse_test.go
@@ -207,8 +207,8 @@ func TestParsePrecedence(t *testing.T) {
 	//   9: AND
 	//  10: OR
 
-	unary := func(op tree.UnaryOperator, expr tree.Expr) tree.Expr {
-		return &tree.UnaryExpr{Operator: op, Expr: expr}
+	unary := func(op tree.UnaryOperatorSymbol, expr tree.Expr) tree.Expr {
+		return &tree.UnaryExpr{Operator: tree.MakeUnaryOperator(op), Expr: expr}
 	}
 	binary := func(op tree.BinaryOperator, left, right tree.Expr) tree.Expr {
 		return &tree.BinaryExpr{Operator: op, Left: left, Right: right}

--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -93,6 +93,21 @@ func processUnaryQualOp(
   op tree.Operator,
   expr tree.Expr,
 ) (tree.Expr, int) {
+  e, code := processUnaryQualOpInternal(sqllex, op, expr)
+  if code != 0 {
+    return e, code
+  }
+  if e, ok := e.(*tree.UnaryExpr); ok {
+    e.Operator.IsOperator = true
+  }
+  return e, code
+}
+
+func processUnaryQualOpInternal(
+  sqllex sqlLexer,
+  op tree.Operator,
+  expr tree.Expr,
+) (tree.Expr, int) {
   switch op := op.(type) {
   case tree.UnaryOperator:
     return &tree.UnaryExpr{Operator: op, Expr: expr}, 0
@@ -108,7 +123,7 @@ func processUnaryQualOp(
   case tree.ComparisonOperator:
     switch op {
     case tree.RegMatch:
-      return &tree.UnaryExpr{Operator: tree.UnaryComplement, Expr: expr}, 0
+      return &tree.UnaryExpr{Operator: tree.MakeUnaryOperator(tree.UnaryComplement), Expr: expr}, 0
     }
   }
   sqllex.Error(fmt.Sprintf("unknown unary operator %s", op))
@@ -10299,15 +10314,15 @@ a_expr:
   }
 | '~' a_expr %prec UMINUS
   {
-    $$.val = &tree.UnaryExpr{Operator: tree.UnaryComplement, Expr: $2.expr()}
+    $$.val = &tree.UnaryExpr{Operator: tree.MakeUnaryOperator(tree.UnaryComplement), Expr: $2.expr()}
   }
 | SQRT a_expr
   {
-    $$.val = &tree.UnaryExpr{Operator: tree.UnarySqrt, Expr: $2.expr()}
+    $$.val = &tree.UnaryExpr{Operator: tree.MakeUnaryOperator(tree.UnarySqrt), Expr: $2.expr()}
   }
 | CBRT a_expr
   {
-    $$.val = &tree.UnaryExpr{Operator: tree.UnaryCbrt, Expr: $2.expr()}
+    $$.val = &tree.UnaryExpr{Operator: tree.MakeUnaryOperator(tree.UnaryCbrt), Expr: $2.expr()}
   }
 | a_expr '+' a_expr
   {
@@ -10676,7 +10691,7 @@ b_expr:
   }
 | '~' b_expr %prec UMINUS
   {
-    $$.val = &tree.UnaryExpr{Operator: tree.UnaryComplement, Expr: $2.expr()}
+    $$.val = &tree.UnaryExpr{Operator: tree.MakeUnaryOperator(tree.UnaryComplement), Expr: $2.expr()}
   }
 | b_expr '+' b_expr
   {
@@ -11583,9 +11598,9 @@ all_op:
 | REGIMATCH { $$.val = tree.RegIMatch }
 | NOT_REGIMATCH { $$.val = tree.NotRegIMatch }
 | AND_AND { $$.val = tree.Overlaps }
-| '~' { $$.val = tree.UnaryComplement }
-| SQRT { $$.val = tree.UnarySqrt }
-| CBRT { $$.val = tree.UnaryCbrt }
+| '~' { $$.val = tree.MakeUnaryOperator(tree.UnaryComplement) }
+| SQRT { $$.val = tree.MakeUnaryOperator(tree.UnarySqrt) }
+| CBRT { $$.val = tree.MakeUnaryOperator(tree.UnaryCbrt) }
 
 operator_op:
   all_op

--- a/pkg/sql/parser/testdata/parse/select_exprs
+++ b/pkg/sql/parser/testdata/parse/select_exprs
@@ -1635,7 +1635,15 @@ SELECT 1 << 2, 1 << 2 -- identifiers removed
 parse
 SELECT OPERATOR(+) 1, OPERATOR(-) 1, OPERATOR(~) 1, OPERATOR(|/) 1, OPERATOR(||/) 1
 ----
-SELECT 1, -1, ~1, |/1, ||/1 -- normalized!
-SELECT (1), (-1), (~(1)), (|/(1)), (||/(1)) -- fully parenthetized
-SELECT _, _, ~_, |/_, ||/_ -- literals removed
-SELECT 1, -1, ~1, |/1, ||/1 -- identifiers removed
+SELECT 1, -1, OPERATOR(~)1, OPERATOR(|/)1, OPERATOR(||/)1 -- normalized!
+SELECT (1), (-1), (OPERATOR(~)(1)), (OPERATOR(|/)(1)), (OPERATOR(||/)(1)) -- fully parenthetized
+SELECT _, _, OPERATOR(~)_, OPERATOR(|/)_, OPERATOR(||/)_ -- literals removed
+SELECT 1, -1, OPERATOR(~)1, OPERATOR(|/)1, OPERATOR(||/)1 -- identifiers removed
+
+parse
+SELECT OPERATOR ( ~ ) 1 COLLATE ident AS ILIKE FROM ident
+----
+SELECT OPERATOR(~)1 COLLATE ident AS ilike FROM ident -- normalized!
+SELECT (OPERATOR(~)((1) COLLATE ident)) AS ilike FROM ident -- fully parenthetized
+SELECT OPERATOR(~)_ COLLATE ident AS ilike FROM ident -- literals removed
+SELECT OPERATOR(~)1 COLLATE ident AS _ FROM _ -- identifiers removed

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -108,7 +108,9 @@ func (*UnaryOp) preferred() bool {
 	return false
 }
 
-func unaryOpFixups(ops map[UnaryOperator]unaryOpOverload) map[UnaryOperator]unaryOpOverload {
+func unaryOpFixups(
+	ops map[UnaryOperatorSymbol]unaryOpOverload,
+) map[UnaryOperatorSymbol]unaryOpOverload {
 	for op, overload := range ops {
 		for i, impl := range overload {
 			casted := impl.(*UnaryOp)
@@ -124,7 +126,7 @@ func unaryOpFixups(ops map[UnaryOperator]unaryOpOverload) map[UnaryOperator]unar
 type unaryOpOverload []overloadImpl
 
 // UnaryOps contains the unary operations indexed by operation type.
-var UnaryOps = unaryOpFixups(map[UnaryOperator]unaryOpOverload{
+var UnaryOps = unaryOpFixups(map[UnaryOperatorSymbol]unaryOpOverload{
 	UnaryMinus: {
 		&UnaryOp{
 			Typ:        types.Int,

--- a/pkg/sql/sem/tree/expr.go
+++ b/pkg/sql/sem/tree/expr.go
@@ -103,7 +103,7 @@ type Operator interface {
 	operator()
 }
 
-var _ Operator = UnaryOperator(0)
+var _ Operator = (*UnaryOperator)(nil)
 var _ Operator = BinaryOperator(0)
 var _ Operator = ComparisonOperator(0)
 
@@ -1254,22 +1254,41 @@ func (node *BinaryExpr) Format(ctx *FmtCtx) {
 	binExprFmtWithParen(ctx, node.Left, node.Operator.String(), node.Right, node.Operator.isPadded())
 }
 
-// UnaryOperator represents a unary operator.
-type UnaryOperator int
+// UnaryOperator represents a unary operator used in a UnaryExpr.
+type UnaryOperator struct {
+	Symbol UnaryOperatorSymbol
+	// IsOperator is true if OPERATOR(symbol) is used.
+	IsOperator bool
+}
+
+// MakeUnaryOperator creates a UnaryOperator given a symbol.
+func MakeUnaryOperator(symbol UnaryOperatorSymbol) UnaryOperator {
+	return UnaryOperator{Symbol: symbol}
+}
+
+func (o UnaryOperator) String() string {
+	if o.IsOperator {
+		return fmt.Sprintf("OPERATOR(%s)", o.Symbol.String())
+	}
+	return o.Symbol.String()
+}
 
 func (UnaryOperator) operator() {}
 
-// UnaryExpr.Operator
+// UnaryOperatorSymbol represents a unary operator.
+type UnaryOperatorSymbol int
+
+// UnaryExpr.Operator.Symbol
 const (
-	UnaryMinus UnaryOperator = iota
+	UnaryMinus UnaryOperatorSymbol = iota
 	UnaryComplement
 	UnarySqrt
 	UnaryCbrt
 
-	NumUnaryOperators
+	NumUnaryOperatorSymbols
 )
 
-var _ = NumUnaryOperators
+var _ = NumUnaryOperatorSymbols
 
 var unaryOpName = [...]string{
 	UnaryMinus:      "-",
@@ -1278,8 +1297,8 @@ var unaryOpName = [...]string{
 	UnaryCbrt:       "||/",
 }
 
-func (i UnaryOperator) String() string {
-	if i < 0 || i > UnaryOperator(len(unaryOpName)-1) {
+func (i UnaryOperatorSymbol) String() string {
+	if i < 0 || i > UnaryOperatorSymbol(len(unaryOpName)-1) {
 		return fmt.Sprintf("UnaryOp(%d)", i)
 	}
 	return unaryOpName[i]
@@ -1303,7 +1322,7 @@ func (node *UnaryExpr) Format(ctx *FmtCtx) {
 	_, isOp := e.(operatorExpr)
 	_, isDatum := e.(Datum)
 	_, isConstant := e.(Constant)
-	if isOp || (node.Operator == UnaryMinus && (isDatum || isConstant)) {
+	if isOp || (node.Operator.Symbol == UnaryMinus && (isDatum || isConstant)) {
 		ctx.WriteByte('(')
 		ctx.FormatNode(e)
 		ctx.WriteByte(')')
@@ -1322,14 +1341,14 @@ func NewTypedUnaryExpr(op UnaryOperator, expr TypedExpr, typ *types.T) *UnaryExp
 	node := &UnaryExpr{Operator: op, Expr: expr}
 	node.typ = typ
 	innerType := expr.ResolvedType()
-	for _, o := range UnaryOps[op] {
+	for _, o := range UnaryOps[op.Symbol] {
 		o := o.(*UnaryOp)
 		if innerType.Equivalent(o.Typ) && node.typ.Equivalent(o.ReturnType) {
 			node.fn = o
 			return node
 		}
 	}
-	panic(errors.AssertionFailedf("invalid TypedExpr with unary op %d: %s", op, expr))
+	panic(errors.AssertionFailedf("invalid TypedExpr with unary op %d: %s", op.Symbol, expr))
 }
 
 // FuncExpr represents a function call.

--- a/pkg/sql/sem/tree/normalize.go
+++ b/pkg/sql/sem/tree/normalize.go
@@ -84,7 +84,7 @@ func (expr *UnaryExpr) normalize(v *NormalizeVisitor) TypedExpr {
 		return val
 	}
 
-	switch expr.Operator {
+	switch expr.Operator.Symbol {
 	case UnaryMinus:
 		// -0 -> 0 (except for float which has negative zero)
 		if val.ResolvedType().Family() != types.FloatFamily && v.isNumericZero(val) {
@@ -104,7 +104,7 @@ func (expr *UnaryExpr) normalize(v *NormalizeVisitor) TypedExpr {
 			}
 		// - (- a) -> a
 		case *UnaryExpr:
-			if b.Operator == UnaryMinus {
+			if b.Operator.Symbol == UnaryMinus {
 				return b.TypedInnerExpr()
 			}
 		}

--- a/pkg/sql/sem/tree/operators_test.go
+++ b/pkg/sql/sem/tree/operators_test.go
@@ -131,7 +131,7 @@ func TestOperatorVolatilityMatchesPostgres(t *testing.T) {
 
 	// Check unary ops. We don't just go through the map so we process them in
 	// an orderly fashion.
-	for op := UnaryOperator(0); op < NumUnaryOperators; op++ {
+	for op := UnaryOperatorSymbol(0); op < NumUnaryOperatorSymbols; op++ {
 		for _, impl := range UnaryOps[op] {
 			o := impl.(*UnaryOp)
 			check(op.String(), nil /* leftType */, o.Typ, o.Volatility)

--- a/pkg/sql/sem/tree/type_check.go
+++ b/pkg/sql/sem/tree/type_check.go
@@ -1323,7 +1323,7 @@ func (expr *Subquery) TypeCheck(_ context.Context, sc *SemaContext, _ *types.T) 
 func (expr *UnaryExpr) TypeCheck(
 	ctx context.Context, semaCtx *SemaContext, desired *types.T,
 ) (TypedExpr, error) {
-	ops := UnaryOps[expr.Operator]
+	ops := UnaryOps[expr.Operator.Symbol]
 
 	typedSubExprs, fns, err := typeCheckOverloadedExprs(ctx, semaCtx, desired, ops, false, expr.Expr)
 	if err != nil {

